### PR TITLE
Verify hooks after collection completes

### LINKED ISSUE #1821: Pytest verifies if hooks are valid before processing "pytest_plugins" variable
Consider:

``` python
# content of plugin.py
class Hooks:
    def pytest_my_hook(self, config):
        pass

def pytest_configure(config):
    config.pluginmanager.add_hookspec(Hooks)

# content of conftest.py
pytest_plugins = ['plugin']

def pytest_my_hook(self, config):
    pass

# content of test_foo.py
def test():
    pass
```

This produces this error:

```
pytest test_foo.py
Traceback (most recent call last):
  File "E:\Miniconda\envs\pytest\Scripts\pytest-script.py", line 9, in <module>
    load_entry_point('pytest==3.0.0.dev1', 'console_scripts', 'pytest')()
  File "c:\users\bruno\pytest\_pytest\config.py", line 56, in main
    config.pluginmanager.check_pending()
  File "c:\users\bruno\pytest\_pytest\vendored_packages\pluggy.py", line 490, in check_pending
    (name, hookimpl.plugin))
_pytest.vendored_packages.pluggy.PluginValidationError: unknown hook 'pytest_my_hook' in plugin <module 'conftest' from 'C:\Users\bruno\pytest\tmp\conftest.py'>
```

It seems pytest is validating hooks before actually processing all hooks. 

This is using the `features` branch just pre-release of 3.0 (be9356aeb).

### PR DESCRIPTION
**Pull request by nicoddemus**
Fix #1821

### CODE REVIEW

**Comment by nicoddemus** [comment]
Strange, Linux is breaking because `pytester` installs `LsofFdLeakChecker` and for some reason it is failing validation... will have to investigate some other time. Meanwhile if someone has any tips, I'm all ears!

**Comment by nicoddemus** [comment]
Oh wow, so the error was that `LsofFdLeakChecker` was declaring `pytest_runtest_item` instead of `pytest_runtest_protocol`. I guess this plugin was not doing anything for a long time now. 😅

**Comment by nicoddemus** [comment]
And not surprisingly some tests now fail because we are now detecting fd leaks again. 😅 

That hookwrapper doesn't seem right as well, it is breaking the entire run instead of failing the offending test.

File: _pytest/config.py [hunk 1]
```diff
@@ -54,7 +54,6 @@ def main(args=None, plugins=None):
             return 4
         else:
             try:
-                config.pluginmanager.check_pending()
```
Comment (RonnyPfannschmidt) [inline _pytest/config.py]: should we remove this check just because we added a new one at a different time in execution?

Comment (nicoddemus) [inline _pytest/config.py]: I think so, that was the main motivation of this PR (at least for me): to delay hook checking until we have loaded all plugins, included those delayed by the `pytest_plugins` mechanism; this avoid having to write [deferred installation](https://docs.pytest.org/en/latest/writing_plugins.html#optionally-using-hooks-from-3rd-party-plugins mechanisms) if the hooks declared by other plugins are required by yours.

Also see #1821 for a detailed example.

Comment (RonnyPfannschmidt) [inline _pytest/config.py]: @nicoddemus oh right, thanks for the reference, my understanding had been wrong

Comment (nicoddemus) [inline _pytest/config.py]: No worries!

Do you or anybody else have any pointers on how to fix the tests leaking FDs?

Comment (RonnyPfannschmidt) [inline _pytest/config.py]: unfortunately not, does it hapen locally as well

Comment (nicoddemus) [inline _pytest/config.py]: Probably, but I don't have a Linux box with me to check this now

Comment (nicoddemus) [inline _pytest/config.py]: Since this is unrelated to this PR (the PR actually fixed the plugin and it is working again), should we create an issue for it and disable the plugin internally?

Comment (RonnyPfannschmidt) [inline _pytest/config.py]: sounds good

**Comment by nicoddemus** [comment]
I think if nobody else wants to chime in, we can close this now.
